docs/how-to-guides/feathr-udfs.md

```diff
@@ -0,0 +1,182 @@
+2. The function should accept only one Spark Dataframe as input, and return one Spark Dataframe as the result. Multiple DataFrames are not supported. There's no limitation on names, and by convention it is usually called `df`, but you can use any name you want.
+3. The "contract" between users and Feathr here is - as long as you give the UDF a dataframe as an input and the output is a dataframe, then Feathr will accept and run this UDF function.
+4. For all the feature definitions, you should use the output of the UDF as the new column name (if you have changed the name of the columns). For example, you have the following code, then you won't be able to use `fare_amount_cents` in your feature definition, since it is not in the output dataframe of this UDF. However, you will be able to use `day_of_week` in this UDF.
+
+```python
+def add_new_dropoff_and_fare_amount_column(df: DataFrame):
+    from pyspark.sql.functions import dayofweek
+    df = df.withColumn("day_of_week", dayofweek("lpep_dropoff_datetime"))
+    df = df.drop("fare_amount_cents")
+    return df
+```
+
+4. Currently, "chained" functions are not supported. I.e. the example below is not supported:
+
+```python
+def multiply_100(input_val):
+  return input_val * 100
+
+def add_new_dropoff_and_fare_amount_column(df: DataFrame):
+    from pyspark.sql.functions import dayofweek
+    df = df.withColumn("f_day_of_week", dayofweek("lpep_dropoff_datetime"))
+    df = df.withColumn("fare_amount_cents",  multiply_100("fare_amount"))
+    return df
+```
+
+Per the first limitation (i.e. the functions should be self-contained), you should consider the following way:
+
+```python
+def add_new_dropoff_and_fare_amount_column(df: DataFrame):
+    from pyspark.sql.functions import dayofweek
+    df = df.withColumn("f_day_of_week", dayofweek("lpep_dropoff_datetime"))
+
+    def multiply_100(input_val):
+      return input_val * 100
+
+    df = df.withColumn("fare_amount_cents",  multiply_100("fare_amount"))
+    return df
+```
+
+### Best Practices
+
+1. One best practice is if you have multiple transformation that you want to define for one input source, you can define multiple UDFs. For example, as shown below, for the same input source, you can define different input UDFs and define completely different features over those two UDFs. Refer to [Illustration on How Feathr UDF works](#illustration-on-how-feathr-udf-works) for an illustration.
```

Comment:
This section looks unfinished as there is only one best practice. We could remove the number bullet point and rephrase senesce before the second one is added.

Reply:
solved